ref(insights): Add platform selector hook to screenloads

narsaynorath opened the pull request:
There's a lot of code around that's specific for determining if a single project is selected and what platform that project is. I'm adding the platform selector to app starts and before doing that I wanted to extract some of that code to simplify the components more. This is in preparation to add the platform selector to App Starts.

This PR:
- Adds a `usePlatformSelector` hook that returns the project if there's a single selection, the platform, whether it's cross platform, and the platform selector component
- Removes a lot of duplicated code and avoids prop drilling project down through the components until it's necessary to do the query.

I also removed any cases of `organization.features.includes('spans-first-ui');` because these modules should already be gated by that feature flag.

narsaynorath committed:
Add platform selector

Uses a hook to pull the current project selection and returns no project
if it's not cross platform. Also returns the platform selector
component. This refactors a bunch of files to remove project from props
and just calls the hook

narsaynorath committed:
Replace old instances of platform with hook usage

narsaynorath committed:
Revert changes to app starts

narsaynorath committed:
Merge branch 'master' into nar/ref/insights-screenloads-make-platform-selector-hook

narsaynorath committed:
Change platform -> selectedPlatform

narsaynorath committed:
wip: tests

narsaynorath committed:
add tests

narsaynorath committed:
Fix imports and simplify

edwardgou-sentry commented on static/app/views/performance/mobile/usePlatformSelector.tsx:
```diff
@@ -0,0 +1,45 @@
+      return null;
+    }
+
+    return projects.find(p => p.id === String(selection.projects));
```
```suggestion
    return projects.find(p => p.id === String(selection.projects)) ?? null;
```

nit: unless we handle `undefined` vs `null` project differently where `usePlatformSelector` is called, I feel like it might be cleaner if we only return either `undefined` or `null`.

edwardgou-sentry commented on static/app/views/performance/mobile/usePlatformSelector.tsx:
```diff
@@ -0,0 +1,45 @@
+    project,
+    selectedPlatform,
+    isProjectCrossPlatform,
+    PlatformSelector,
```
question: how come we return `PlatformSelector` here? Doesn't seem like we make any changes to the component here, so I think we can make this hook lighter by expecting users to import this component on their own.

narsaynorath replied:
Yeah, I thought this was a bit odd but I figured it was okay since this hook was also returning the selected platform, so it made sense to return the component that modified the selection. I looked it up and it's supposed to be pretty anti-pattern-y because it doesn't allow for flexibility so I took it out and renamed the hook to `useCrossPlatformProject`. It's not the best name but I'm unsure of what to call it for now

edwardgou-sentry commented on static/app/views/performance/mobile/usePlatformSelector.tsx:
```diff
@@ -0,0 +1,45 @@
+  }, [projects, selection.projects]);
+
+  const isProjectCrossPlatform = useMemo(
+    () => (project ? isCrossPlatform(project) : false),
```
```suggestion
    () => !!(project && isCrossPlatform(project)),
```
nit: similar feedback, if we don't have expect logic on `undefined` values, maybe we can just force this too a boolean

edwardgou-sentry reviewed (approved):
looks reasonable to me! just left some nits and a question

narsaynorath committed:
CR fixes

- prevent undefined from passing as project
- remove component returned from hook
- rename hook
